### PR TITLE
StinkingBanana_Keyword_Count

### PR DESCRIPTION
**Pull request by wei0831**
## Adding (keywordcount) variable to use in Keywords' response.

**scripts\commands\customCommands.js**
-  Modified function tags

**scripts\handlers\keywordHandler.js**
- Modified event ircChannelMessage

**web\panel\js\keywordsPanel.js**
- Modified function onMessage, doQuery, addKeywordnew, updateKeyword, editKeywordnew
- Added function incrKeywordCount, decrKeywordCount, resetKeywordCount
- Export  $.incrKeywordCount,  $.decrKeywordCount, $.resetKeywordCount

**web\panel\keywords.html**
- Added an input field for Keyword Count in the modal

## The result will give you the following features:
keyword_count
- The keywords web panel will show Count, Keyword, and Response
and three buttons on the left for each keyword:
[+] button: to add count by 1
[-] button: to decrease count by 1
[reset] button: to reset count to 0

keyword_count1
- When you try to edit keyword in the kewords web panel, the keywords modal will have one extra input field allowing you to set the count for the keyword.

keyword_count2
- When you try to add keyword in the kewords web panel, the keywords modal will have one extra input field allowing you to set the initial count for the keyword.

- You can use (keywordcount) in the keyword response to show count.

## Example
Keywowrd: ```^[hH]ello.*```
Regex: ```On```
Response: ```What's up? [+(keywordcount)]```
Usage:
```
StinkingBanana: Hello my friend :D
PhantomBot: What's up?  [+1234]
StinkingBanana: Hello Again :D
PhantomBot: What's up?  [+1235]
```
---
[Original Post](https://community.phantombot.tv/t/keyword-count/3674) @ community.phantombot.tv

### CODE REVIEW

File: javascript-source/commands/customCommands.js [hunk 1]
```diff
@@ -203,6 +203,21 @@
...
         }
 
+        if (message.match(/\(keywordcount\s(.+)\)/g)) {
+            var input_keyword = message.match(/\(keywordcount\s(.+)\)/)[1],
+                keyword_info = JSON.parse($.inidb.get('keywords', input_keyword));
+        
+            if('count' in keyword_info){
+                ++keyword_info["count"];
+            }
+            else{
```
Comment (IllusionaryOne) [inline javascript-source/commands/customCommands.js]: Do hug the brace and put a space.  i.e.: } else {

File: javascript-source/commands/customCommands.js [hunk 2]
```diff
@@ -203,6 +203,21 @@
             message = $.replace(message, '(count)', $.inidb.get('commandCount', event.getCommand()));
         }
 
+        if (message.match(/\(keywordcount\s(.+)\)/g)) {
+            var input_keyword = message.match(/\(keywordcount\s(.+)\)/)[1],
+                keyword_info = JSON.parse($.inidb.get('keywords', input_keyword));
+        
+            if('count' in keyword_info){
```
Comment (IllusionaryOne) [inline javascript-source/commands/customCommands.js]: Please keep the same code style as above.  Space after if (), space before braces.

**Comment by IllusionaryOne** [comment]
The test results look good, the code seems sane.  I did put in some comments for cleanup.

@ScaniaTV to you to review.

**Review by IllusionaryOne — COMMENTED** [review]
A couple of items to cleanup.

File: javascript-source/commands/customCommands.js (hunk 2 above)
Comment (wei0831) [reply javascript-source/commands/customCommands.js]: Thanks so much for your feedback. I have updated accordingly.

File: javascript-source/commands/customCommands.js (hunk 1 above)
Comment (wei0831) [reply javascript-source/commands/customCommands.js]: Thanks so much for your feedback. I have updated accordingly.

**Comment by wei0831** [comment]
I have squashed commits into one commit, and rebased on top of the latest version of upstream master.

**Comment by BrandenB** [comment]
I like this idea, but I don't think the panel buttons are necessary, same with the count showing in the table. It will probably just confuse users and changes the keyword system into a counter.

**Comment by wei0831** [comment]
I understand. 
I have been using this keyword count for a while now, and I have not found the need to use the panel buttons and modal count field. I put those there just in case if I needed to reset the count.

I thought of just using commands to set and reset counts.
eg.
```
User: !kcount get regex:^[hH]ello.*
Bot:  This keyword has been used 1234 times.
---
User: !kcount reset regex:^[hH]ello.*
User: !kcount get regex:^[hH]ello.*
Bot:  This keyword has been used 0 times.
---
User: !kcount set regex:^[hH]ello.* 100
User: !kcount get regex:^[hH]ello.*
Bot:  This keyword has been used 100 times.
```

However, typing out the keyword trigger phrase is kinda crazy. So, I did not go this route.
I could not figure out a better idea just yet. 
Please let me know if there is a better way doing this, I will do it.

**Comment by wei0831** [comment]
Beside keyword count, I have several other ideas.
1. Making keyword system to allow multiple keywords to trigger in one sentence.
eg.
```
Keyword:  regex:^[hH]ello.*
Response: [SoundEffect] Hello (playsound hello)
Keyword:  regex:.*my friend.*
Response: [SoundEffect] my friend (playsound my_friend)
---
User: Hello my friend
Bot: [SoundEffect] Hello
Bot: [SoundEffect] my friend
```
In this particular example, the audiences can combine keywords to form interesting sentence and mimic text to speech.
You might wonder why not just use speechchat for the text to speech.
It's different. Cause I get to choose what keyword can trigger the speech with what sound effect.

2. Keyword can have more than one responses, and can choose to reply using random response.
eg.
```
# Random Responses
Keyword:  regex:^[hH]ello.*
Response1: What's up?
Response2: How are you doing?
Response3: Welcome to my channel :D
---
User1: Hello my friend
Bot: What's up?
User2: Hello sir?
Bot: Welcome to my channel :D
```

I am new to the PhantomBot community. I might missed some rules.
I hope this will not cause too much inconvenience to you.

**Comment by BrandenB** [comment]
I can understand having a command to reset the count is kinda pointless for keywords since we allow regex. However I would be fine with a button next to the current ones to reset it, if it has a tool-tip saying what it does that is.

As for multiple responses for keywords, or allowing multiple keywords being used in a message, I cannot allow that in the build. Reason being - Twitch only allows so many messages to be sent in a certain time frame. As of right now, the bot is limited to 20 messages in 30 seconds, timeouts are included into this limit, even though we allow 100 to be sent since that's the channel moderator limit. Once the bot has sent 20 messages within 30 seconds it will stop sending until the 30 seconds has passed; this is also why we don't allow commands to have multiple messages either.

For random responses, the user can use the `(randfilerand)` variable and read a random line from a text file. If you have a better way and more user-friendly way of doing go for it!

I like all of your ideas but sadly Twitch prevents us from adding cool features at times.

**Comment by wei0831** [comment]
Thanks so much for your feedback :D Very helpful.

I will revise the PR right away.

## The revised version:
keywordsPanel_revised
- The keywords web panel will show Keyword Count with a Reset button 
- You can use (keywordcount) in the keyword response to increment count and show count.

Is the revised version looking okay?
or do you want not having the keyword count field and move the reset button next to delete button?
like the following:
keywordsPanel_revised_option

**Comment by wei0831** [comment]
## The revised version:
pr1
- The keywords web panel will show Keyword Count with a Reset button 
- You can use (keywordcount) in the keyword response to show count.

## Example
Keywowrd: ```^[hH]ello.*```
Regex: ```On```
Response: ```What's up? [+(keywordcount)]```
Usage:
```
StinkingBanana: Hello my friend :D
PhantomBot: What's up?  [+1234]
StinkingBanana: Hello Again :D
PhantomBot: What's up?  [+1235]
```

**Comment by BrandenB** [comment]
I'm fine with both of them. Choose the one you like best and I'll merge.

**Comment by wei0831** [comment]
Sweet. I like the current one. You can merge it :D
Thanks so much.

**Comment by CreativeCodeCat** [comment]
By the way @ScaniaTV this does not work. when you first add a keyword it has no count in the database until you click the reset counter button and running keyword dont make a different in making the counter going up or does it REQUIRE to have (keywordcount) in the keyword?

video clip of proof http://alixe.pro/share/file/205413981X150318.mp4

So i worked out how it works but its very miss leading in the dashboard 

img is here

**Comment by wei0831** [comment]
Yes. You are correct, initially it does not include the count in the keyword attribute.
Since I removed the input field from the modal, it no longer add the count when creating new keyword. (My first version will work more like you expected.)

Now, the only check is in the customCommands.js
```
if ('count' in keyword_info) {
    ++keyword_info["count"];
} else {
    keyword_info["count"] = 1;
}
```
When the ```(keywordcount)``` is used, if the count attribute is not present, it will add it.
So, the logic is if you don't include ```(keywordcount)``` in the keyword response, it will never get increased.

As for the panel's  keyword count display, you will need to click the refresh button on the top right corner to see the current count.

**Comment by CreativeCodeCat** [comment]
this would be better if the counter went up regardless if the keyword has (keywordcount) in it that way u can still see it in the dashboard even if the keyword doesn't tell you also one key word shouldn't be able to edit another counter thats just ridiculous

**Comment by wei0831** [comment]
```(keywordcount)``` is mimicking how ```(count)``` works. 
If you don't use ```(count)``` in the commands response, the count will never get increased.

I am guessing the original idea was to give users the options to decide whether to count or not.
Thus, I made the ```(keywordcount)``` behaves the same as ```(count)```.

I don't understand this part: "one keyword shouldn't be able to edit another counter".

Anyway, I am pretty open to any better ideas. I will make it right and I would love to make it better :D

**Comment by CreativeCodeCat** [comment]
check the image.... i was doing the keyword THIS but the counter for TEST was going up not the one for THIS

**Comment by wei0831** [comment]
I see.

Here you go
Edit: (better gif)
keywordcount

**Comment by CreativeCodeCat** [comment]
Thats not my point... i was running THIS and it did not change that at all it was changing the counter for THAT and THAT didnt have a (keywordcount) at allso  shouldnt happen also if you add 2 (keywordcount)'s to the same keywrord it error's which also shouldnt happen 

`[03-15-2018 @ 21:11:32.597 GMT] [ERROR] [init.js:308] (hook.call, ircChannelMessage, ./handlers/keywordHandler.js) TypeError: Can't use 'in' on a non-object.`

**Comment by wei0831** [comment]
I cannot reproduce.
```
[03-15-2018 @ 21:11:32.597 GMT] [ERROR] [init.js:308] (hook.call, ircChannelMessage, ./handlers/keywordHandler.js) TypeError: Can't use 'in' on a non-object.
```

But, here is the gif with console window
keywordcount2

**Comment by wei0831** [comment]
All right. 
I have created a new PR #1918 solving the problems.

**Pull request merged**